### PR TITLE
Add necessary fields to RelationConfig

### PR DESCRIPTION
**Pull request by colin-rogers-dbt**
- update RelationConfig to include all 1p adapter expected fields

resolves #
[docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose) dbt-labs/docs.getdbt.com/#

### Checklist

- [ ] I have read [the contributing guide](https://github.com/dbt-labs/dbt-adapter/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [ ] I have run this code in development and it appears to resolve the stated issue
- [ ] This PR includes tests, or tests are not required/relevant for this PR
- [ ] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX

### CODE REVIEW

File: dbt/adapters/contracts/relation.py [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
-from typing import Dict, Optional
+from dbt_common.contracts.config.materialization import OnConfigurationChangeOption
+from typing import Dict, Optional, Any, Union, List
 
 from dbt_common.contracts.util import Replaceable
 from dbt_common.dataclass_schema import StrEnum, dbtClassMixin
 from dbt_common.exceptions import CompilationError, DataclassNotDictError
 from dbt_common.utils import deep_merge
-from typing_extensions import Protocol
+from typing_extensions import Protocol, TypedDict
```
Comment (MichelleArk) [inline dbt/adapters/contracts/relation.py:12]: is this used anywhere?

File: dbt/adapters/contracts/relation.py [hunk 2]
```diff
@@ -18,13 +21,39 @@ class RelationType(StrEnum):
...
+
 class RelationConfig(Protocol):
     name: str
     database: str
     schema: str
     identifier: str
+    compiled_code: Optional[str]
     quoting_dict: Dict[str, bool]
-    config: Dict[str, str]
+    config: MaterializationConfig
```
Comment (mikealfare) [inline dbt/adapters/contracts/relation.py:56]: I don't think we should have a materialization config as an attribute of a relation config. Relations are components of a materialization, but the component matters (e.g. existing, target, temp, backup, etc.). Also, relations exist regardless of a materialization. So at very least, this needs to be optional.

File: dbt/adapters/contracts/relation.py [hunk 3]
```diff
@@ -1,12 +1,15 @@
+from abc import ABC
```
Comment (mikealfare) [inline dbt/adapters/contracts/relation.py:1]: Minor nit, these imports are out of order. `dbt_common.contracts.config.materialization` is 3rd party, should be moved down before `dbt_common.contracts.util`. The extra line on line 2 is unnecessary. The imports for `typing` are not alphabetical.

File: dbt/adapters/contracts/relation.py (hunk 2 above)
Comment (colin-rogers-dbt) [reply dbt/adapters/contracts/relation.py:56]: Yea this is confusing I was trying to figure out what RelationConfig.config actually represented and it seems to be the configuration for how an adapter should actually instantiate the resulting database object.

Maybe `RelationGenerationConfig`?

Comment (mikealfare) [reply dbt/adapters/contracts/relation.py:57]: Per our offline conversation, I'm fine leaving this as is.

**Review by mikealfare — APPROVED** [review]

**Pull request merged**